nodejs/aws-infra/experimental/clusterAutoScaling.ts

```diff
@@ -0,0 +1,572 @@
+                args: ClusterAutoScalingLaunchConfigurationArgs,
+                opts: pulumi.ComponentResourceOptions = {}) {
+        super("aws-infra.x.ClusterAutoScalingLaunchConfiguration", name, args, opts);
+
+        const parentOpts = { parent: this };
+
+        const cluster = args.cluster;
+
+        // Create the full name of our CloudFormation stack here explicitly. Since the CFN stack
+        // references the launch configuration and vice-versa, we use this to break the cycle.
+        // TODO[pulumi/pulumi#381]: Creating an S3 bucket is an inelegant way to get a durable,
+        // unique name.
+        const stackName = pulumi.output(args.stackName).apply(sn => sn || new aws.s3.Bucket(name, {}, parentOpts).id);
+
+        // Use the instance provided, or create a new one.
+        const instanceProfile = getInstanceProfile(name, args, parentOpts);
+        const fileSystem = args.fileSystem;
+
+        const securityGroups =
+            pulumi.output(args.securityGroups).apply(g => g || cluster.instanceSecurityGroups.map(g => g.id));
+        const instance = new aws.ec2.LaunchConfiguration(name, {
+            ...args,
+            securityGroups,
+            imageId: getEcsAmiId(args.ecsOptimizedAMIName),
+            instanceType: pulumi.output(args.instanceType).apply(t => t || "t2.micro"),
+            iamInstanceProfile: instanceProfile.id,
+            enableMonitoring: pulumi.output(args.enableMonitoring).apply(b => b !== undefined ? b : true),
+            placementTenancy: pulumi.output(args.placementTenancy).apply(t => t || "default"),
+            rootBlockDevice: pulumi.output(args.rootBlockDevice).apply(d => d || defaultRootBlockDevice),
+            ebsBlockDevices: pulumi.output(args.ebsBlockDevices).apply(d => d || defaultEbsBlockDevices),
+            userData: getInstanceUserData(cluster, args, stackName),
+        }, parentOpts);
+
+        this.instance = instance;
+        this.cluster = cluster;
+        this.stackName = stackName;
+        this.instanceProfile = instanceProfile;
+        this.fileSystem = fileSystem;
+
+        this.registerOutputs({
+            instance,
+            cluster,
+            stackName,
+            instanceProfile,
+            fileSystem,
+        });
+    }
+
+    public static defaultInstanceProfilePolicyDocument(): aws.iam.PolicyDocument {
+        return {
+            Version: "2012-10-17",
+            Statement: [{
+                Action: [
+                    "sts:AssumeRole",
+                ],
+                Effect: "Allow",
+                Principal: {
+                    Service: [ "ec2.amazonaws.com" ],
+                },
+            }],
+        };
+    }
+
+    public static defaultInstanceProfilePolicyARNs() {
+        return [aws.iam.AmazonEC2ContainerServiceforEC2Role, aws.iam.AmazonEC2ReadOnlyAccess];
+    }
+}
+
+(<any>ClusterAutoScalingLaunchConfiguration).doNotCapture = true;
+
+const defaultRootBlockDevice = {
```

Comment:
Any reason we need to define these, rather than just relying on the provider's default?

Reply:
We don't have to.  But then our awsinfra.Cluster story is a lot weaker than our cloud.Service story.  specifically, if you wanted to move down from cloud to awsinfra, you'd have to figure out all these defaults and add them in.  A high level goal of awsinfra that @lukehoban and i are exploring is:

> Can we provide sensible starting defaults that allow users to not have to configure so much stuff, while also being flexible enough to ensure we're not boxing users out from having control when tehy need it.

Note; if we did not provide these defaults, then honestly there's not that much value this PR is providing.  After all, the user could just manually make the launch-config themselves, as well as the cloud-formation-stack, as well as the cluster.  But that's not a very pleasant story (especially when you're moving downwards from 'cloud').